lib/Dialect/TfheRustBool/IR/TfheRustBoolAttributes.td

```diff
@@ -0,0 +1,31 @@
+#ifndef LIB_DIALECT_TFHERUSTBOOL_IR_TFHERUSTBOOLATTRIBUTES_TD_
+#define LIB_DIALECT_TFHERUSTBOOL_IR_TFHERUSTBOOLATTRIBUTES_TD_
+
+include "lib/Dialect/TfheRustBool/IR/TfheRustBoolDialect.td"
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/IR/BuiltinDialect.td"
+include "mlir/IR/CommonAttrConstraints.td"
+
+class TfheRustBool_Attr<string name, string attrMnemonic, list<Trait> traits = []>
+    : AttrDef<TfheRustBool_Dialect, name, traits> {
+  let mnemonic = attrMnemonic;
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+def TfheRustBoolGate_Attr : TfheRustBool_Attr<"TfheRustBoolGate", "tfhe_rust_bool_gate"> {
+  let summary = "An Attribute containing an array of strings to store bool gates";
+
+  let description = [{
+    Attribute to store what kind of Boolean gate we are working on
+  }];
+
+  let parameters = (ins
+      ArrayRefParameter<"mlir::StringAttr">: $gate);
```

Comment:
Just realized you could add a Verification method to check whether the string is one of the ones you expect 
https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/#verification

If you share the same code with CGGI (actually, can this gate attribute just be a common attribute between the two?), then the Default typeswitched case above would actually be an unreachable since verification would catch the bad gate string

Reply:
I'm not sure what Asra means by sharing code here (CGGI does not have a packed gate op?), but for verification, you could consider changing this to an ArrayRefParameter of an enum type. You can see how this is done here: https://github.com/google/heir/blob/114e8c3d564273d60e36d398f7a795eca60d4bcd/lib/Dialect/Random/IR/RandomEnums.td#L11-L16

The tabelgen definition linked above creates a vanilla C++ enum class (not actually an attribute), and you can just `ArrayRefParameter<"mlir::heir::foo::YourEnumName">` to put it on an attribute or op.

Reply:
Oh, I meant sharing the verifier for an array of gates with the CGGI gate attribute.
https://github.com/google/heir/blob/114e8c3d564273d60e36d398f7a795eca60d4bcd/lib/Dialect/CGGI/IR/CGGIAttributes.td#L33

(As in, this gate attribute should just be used in both the cggi packed op and the tfhe one. Here's the CGGI packed op: https://github.com/google/heir/blob/114e8c3d564273d60e36d398f7a795eca60d4bcd/lib/Dialect/CGGI/IR/CGGIOps.td#L62

Reply:
I stand corrected! I feel having an enum (which does not need to be attached to any dialect) would avoid the need for a verifier entirely. Then both ops can use ArrayRefParameter<enum> and we can cut down on tablegen code.